### PR TITLE
Prebid Core : add setScriptAttributes to adloader in Utils

### PR DESCRIPTION
**Pull request by spotxslagle**
## Type of change

- [ ] Bugfix
- [ X ] Feature
- [ ] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change
Minor feature
Modification to allow spotxBidAdapter to load its script with attributes, removing the need for appendChild

- contact email of the adapter’s maintainer
- aslagle@magnite.com

@npeceniak @robertrmartinez

### CODE REVIEW

**Comment by ChrisHuie** [comment]
@spotxslagle @robertrmartinez @osazos added 2 reviewers here because of the utils change. Would this be better as 2 separate prs? One to core and one for spotx bid adapter?

**Comment by bretg** [comment]
@spotxslagle - can we assume this obsoletes https://github.com/prebid/Prebid.js/pull/8536 ?

**Comment by spotxslagle** [comment]
@ChrisHuie I think that would be reasonable. I'll refactor it into separate pull requests.  And this change would make the other obsolete.

**Comment by spotxslagle** [comment]
@ChrisHuie It should be ready for review now

**Comment by patmmccann** [comment]
I'm confused here, it seems like there are no changes to the spotx adapter and the appendchild still occurs?

**Comment by patmmccann** [comment]
https://github.com/spotxslagle/Prebid.js/blob/ef8f9f08b5ddfb2463133740813712e10ef58467/modules/spotxBidAdapter.js#L504

**Comment by spotxslagle** [comment]
I was asked if it made sense to add the prebid core change in one pr, and then have a separate pr for SpotX to make use of the change. Since nothing about the core change can be taken advantage of by anyone, it made sense to me to have it separate from a pr fixing an issue with the SpotX adapter. Once this pr is merged I'll have another pr up within a few days

**Review by robertrmartinez — APPROVED** [review]
Looks good, connects to the other PR @dgirardi  was looking into I believe if you want to check this one as well.

File: src/adloader.js [hunk 1]
```diff
@@ -107,6 +109,10 @@ export function loadExternalScript(url, moduleCode, callback, doc) {
 
     jptScript.src = tagSrc;
 
+    if (attributes) {
+      setScriptAttributes(jptScript, attributes);
```
[the remark this answers is not in the document]
Comment (spotxslagle) [reply src/adloader.js:113]: It is needed. Those attributes need to be set before the script is attached to the DOM. I'll investigate the cache key

**Comment by dgirardi** [comment]
In that case I think it's also fine to bypass the cache when attributes are
passed in.

File: src/adloader.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (spotxslagle) [reply src/adloader.js:113]: @dgirardi RE: caching attributes. If a script is cached, loadExternalScript will return that script immediately upon retrieving it from the cache. The setAttributes function is only ever called on the novel script generated inside the requestResource function, which is never reached when a resource has already been cached. So, to answer your question, multiple calls will always result in only the first script tag, and thus the first set of attributes, being returned. It would also be undesirable to include these attributes on the cache key, as they will already be present on the script tag and would both be unused and a violation of the single source of truth principle.

TLDR: Caching attributes would be redundant and the pr as written will never result in attributes being overwritten.

**Comment by spotxslagle** [comment]
Great! Is there anything else someone would like me to address before this is merged? I don't have merge permissions, so someone else will need to push the button

File: src/adloader.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (spotxslagle) [reply src/adloader.js:113]: I've been told in the past that I can be a bit blunt, which doesn't always come off the way I intend. So I want to flag that I am enjoying engaging in the topic of how this feature should be architected.

As it stands, if the same url is requested multiple times, Prebid basically stiff arms the second request and returns the original tag without modification. This includes dropping any callbacks from the second request. This could be considered a bug. 

For the purposes of the SpotX adapter, though, changing the attributes for a tag that already has attributes would not work. I'm not sure if it would break something or if it would just be ignored, but I do know it wouldn't work. So for the purposes for which this feature is being added, the current implementation is the desired one, if not necessarily "proper". 

I would also say that I think it is reasonable to say that a resource can only be loaded once, and any other requests for the same resource will simply go to the cache. Whether this is appropriate or whether it makes more sense to continually modify the request's callbacks or attributes is use case dependent. Given that the current architecture seems to not have raised any concerns so far, I would say that the simplest, and therefore probably best response to avoiding a hard to find bug would be adding a log warning that the resource has already been requested and won't be modified.

As for docs, this function is not present in Prebid's external facing api, and not currently documented. No documentation will be changed as a result of this pr.

**Comment by spotxslagle** [comment]
I modified the JSDocs to note that only the attributes attached to the first request will be added. No other documentation change is needed.

File: src/adloader.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (dgirardi) [reply src/adloader.js:113]: > As it stands, if the same url is requested multiple times, Prebid basically stiff arms the second request and returns the original tag without modification. This includes dropping any callbacks from the second request. 

I don't think the last bit is true (but I haven't tested), from reading the code I think the callback will run. I agree that if it doesn't, it looks like a bug to me - but I don't think it'd be in the scope of this PR (and after such a long time, it is now a feature :)

> For the purposes of the SpotX adapter, though, changing the attributes for a tag that already has attributes would not work.

But that's not what I'm proposing, which is creating a new script element (because the cache is either missing or bypassed). Basically if the attributes are different, treat it as a different resource. It's similar to what the implementation does with the `doc` argument, which was recently added.

> As for docs, this function is not present in Prebid's external facing api, and not currently documented. 

What I mean is, note it in the param documentation, so that if someone gets bit by this I can blame them instead of you :) I saw that you did that while I was typing this, that's good enough for me.

**Comment by spotxslagle** [comment]
@robertrmartinez Do you know who I should be pinging to get this merged?
